File: plot_playground/storytelling/simple_line_date_series_plot.py
"""
A module that handles a plot of a polygonal line that makes only
specific line elements stand out.
"""

import numpy as np

from plot_playground.common import d3_helper
from plot_playground.common import data_helper

PATH_CSS_TEMPLATE = 'storytelling/simple_line_date_series_plot.css'
PATH_JS_TEMPLATE = 'storytelling/simple_line_date_series_plot.js'


def display_plot(
        df,
        date_column,
        normal_columns,
        stands_out_columns,
        width=600,
        height=372,
        title='',
        title_color='#6bb2f8',
        title_font_size=25,
        description='',
        description_font_size=14,
        description_color='#999999',
        y_axis_label='',
        y_axis_prefix='',
        y_axis_suffix='',
        axis_line_color='#999999',
        axis_text_color='#999999',
        axis_font_size=14,
        legend_font_size=14,
        legend_color='#999999',
        stands_out_legend_font_size=14,
        stands_out_legend_font_color='#6bb2f8',
        stands_out_legend_font_weight='bold',
        line_color='#e8e8e8',
        line_size=2.5,
        stands_out_line_color='#acd5ff',
        stands_out_line_size=4.0,
        outer_margin=20,
        x_ticks=5,
        y_ticks=5,
        plot_background_color='#ffffff',
        plot_margin_left=0,
        outer_border_size=1,
        outer_border_color='#cccccc',
        font_family='-apple-system, BlinkMacSystemFont, "Helvetica Neue", YuGothic, "ヒラギノ角ゴ ProN W3", Hiragino Kaku Gothic ProN, Arial, "メイリオ", Meiryo, sans-serif',
        svg_id='',
    ):
    """
    Display a simple line plot on Jupyter. Only particular ones
    are markedly visible.

    See Also
    --------
    https://nbviewer.jupyter.org/github/simon-ritchie/plot_playground/blob/master/documents/storytelling_simple_line_date_series_plot/document.html
        Document of this plot.

    Parameters
    ----------
    df : pandas.DataFrame
        Data frame to be plotted. A date column is required.
    date_column : str
        Column name of the date in the data frame.
    normal_columns : list of str
        A list of the column names of targets for which inconspicuous
        colors are set.
    stands_out_columns : list of str
        A list of column names for which prominent colors are set.
    width : int, default 600
        Width of the plot.
    height : int default 372
        Height of the plot.
    title : str, default ''
        The title of the plot. If an empty character is specified,
        the title is not displayed.
    title_color : str, default '#6bb2f8'
        Title color setting.
    title_font_size : int, defaul 25
        Title font size.
    description : str, default ''
        Explanatory text. It is displayed under the title. It is not
        displayed when an empty character is specified.
    description_font_size : int, default 14
        The font size of description.
    description_color : str, default '#999999'
        The color setting of description.
    y_axis_label : str, default ''
        The label to set on the y axis. It is displayed in a rotated state.
    y_axis_prefix : str, default ''
        A string to set before the value of the y axis.
        e.g., $.
    y_axis_suffix : str, default ''
        A character string to be set after the value of the y axis.
        e.g., %.
    axis_line_color : str, default '#999999'
        Color of axis line.
    axis_text_color : str, defaul '#999999'
        The font color of the axis.
    axis_font_size : int, default 14
        The font size of the axis.
    legend_font_size : int, default 14
        The font size of the legend.
    legend_color : str, default '#999999'
        The color setting of legend.
    stands_out_legend_font_size : int, default 14
        The legend's font size of the place to stand out.
    stands_out_legend_font_color : str, default '#6bb2f8'
        The legend's font color of the place to stand out.
    stands_out_legend_font_weight : str, default 'bold'
        The legend's font weight of the place to stand out.
    line_color : str, default '#e8e8e8'
        Line color of a normal polygonal line.
    line_size : float, default 2.5
        The size of a normal polygonal line.
    stands_out_line_color : str, default '#acd5ff'
        Line color of the line to make it stand out.
    stands_out_line_size : float, default 4.0
        Size of the line to stand out.
    outer_margin : int, default 20
        Edge margin of plot area.
    x_ticks : int, default 5
        Number of steps on the x axis. This number roughly varies depending
        on the value of surplus etc.
    y_ticks : int
        Number of steps in the y axis. This number roughly varies depending
        on the value of surplus etc.
    plot_background_color : str
        The background color of the plot.
    plot_margin_left : int, default 0
        The left margin of the SVG area.
    outer_border_size : int, default 1
        Plot outside border size.
    outer_border_color : str, default '#cccccc'
        Plot outside border color.
    font_family : str
        Font setting. e.g., Meiryo, sans-serif
    svg_id : str, default ''
        ID to set for SVG element. When an empty value is specified,
        a unique character string is generated and used.

    Returns
    -------
    plot_meta : plot_playground.common.d3_helper.PlotMeta
        An object that stores the metadata of the plot.
    """
    if svg_id == '':
        svg_id = d3_helper.make_svg_id()
    css_template_str = d3_helper.read_template_str(
        template_file_path=PATH_CSS_TEMPLATE)
    css_param = {
        'svg_id': svg_id,
        'svg_border_size': outer_border_size,
        'svg_border_color': outer_border_color,
        'font_family': font_family,
        'title_color': title_color,
        'title_font_size': title_font_size,
        'description_font_size': description_font_size,
        'description_color': description_color,
        'legend_font_size': legend_font_size,
        'legend_color': legend_color,
        'stands_out_legend_font_size': stands_out_legend_font_size,
        'stands_out_legend_font_color': stands_out_legend_font_color,
        'stands_out_legend_font_weight': stands_out_legend_font_weight,
        'axis_stroke_color': axis_line_color,
        'axis_text_color': axis_text_color,
        'axis_font_size': axis_font_size,
        'line_color': line_color,
        'line_size': line_size,
        'stands_out_line_color': stands_out_line_color,
        'stands_out_line_size': stands_out_line_size,
    }
    css_template_str = d3_helper.apply_css_param_to_template(
        css_template_str=css_template_str,
        css_param=css_param
    )

    df = df.copy()
    _validate_df_columns(
        df=df, date_column=date_column, normal_columns=normal_columns,
        stands_out_columns=stands_out_columns)
    df = data_helper.cast_df_column_to_date_str(
        df=df, column_name=date_column)
    merged_column_list = [*normal_columns, *stands_out_columns]
    data_helper.validate_null_value_not_exists_in_df(
        df=df, columns=merged_column_list)
    data_helper.validate_all_values_are_numeric(
        df=df, columns=merged_column_list)
    dataset_df = df.loc[:, [date_column, *merged_column_list]]
    dataset = dataset_df.to_dict(orient='record')
    legend_dataset = _make_legend_dataset(
        df=df, date_column=date_column, normal_columns=normal_columns,
        stands_out_columns=stands_out_columns)
    year_str_list = _make_year_str_list(df=df, date_column=date_column)
    y_axis_min = data_helper.get_df_min_value(
        df=df, columns=merged_column_list)
    y_axis_min = min(0, y_axis_min)
    y_axis_max = data_helper.get_df_max_value(
        df=df, columns=merged_column_list)
    js_param = {
        'svg_id': svg_id,
        'svg_width': width,
        'svg_height': height,
        'svg_background_color': plot_background_color,
        'svg_margin_left': plot_margin_left,
        'outer_margin': outer_margin,
        'x_ticks': x_ticks,
        'y_ticks': y_ticks,
        'y_axis_prefix': y_axis_prefix,
        'y_axis_suffix': y_axis_suffix,
        'plot_title': title,
        'plot_description': description,
        'dataset': dataset,
        'date_column': date_column,
        'column_list': normal_columns,
        'stands_out_column_list': stands_out_columns,
        'legend_dataset': legend_dataset,
        'year_str_list': year_str_list,
        'y_axis_min': y_axis_min,
        'y_axis_max': y_axis_max,
        'y_axis_label': y_axis_label,
        'x_axis_min': df[date_column].min(),
        'x_axis_max': df[date_column].max(),
    }
    js_template_str = d3_helper.read_template_str(
        template_file_path=PATH_JS_TEMPLATE)
    js_template_str = d3_helper.apply_js_param_to_template(
        js_template_str=js_template_str,
        js_param=js_param
    )
    html_str = d3_helper.exec_d3_js_script_on_jupyter(
        js_script=js_template_str,
        css_str=css_template_str,
        svg_id=svg_id,
        svg_width=width,
        svg_height=height)

    plot_meta = d3_helper.PlotMeta(
        html_str=html_str,
        js_template_str=js_template_str,
        js_param=js_param,
        css_template_str=css_template_str,
        css_param=css_param)
    return plot_meta


def _make_year_str_list(df, date_column):
    """
    Generate a list containing year strings to be used on the x axis.

    Parameters
    ----------
    df : DataFrame
        A data frame containing a column of date string.
    date_column : str
        Column name of the date.

    Returns
    -------
    year_str_list : list of str
        A list containing characters at the beginning of the year.
        e.g., ["2018-01-01", "2019-01-01"]
    """
    min_date_str = df[date_column].min()
    date_unique_arr = df[date_column].unique()
    year_str_list = []
    for date_str in date_unique_arr:
        year_str = data_helper.get_year_str_from_date_str(
            date_str=date_str)
        year_str_list.append(year_str)
    year_str_list = np.unique(year_str_list).tolist()
    year_str_list.sort()
    for i, year_str in enumerate(year_str_list):
        year_str += '-01-01'
        if year_str < min_date_str:
            year_str = min_date_str
        year_str_list[i] = year_str
    return year_str_list


def _make_legend_dataset(
        df, date_column, normal_columns, stands_out_columns):
    """
    Make a data set for the legend.

    Parameters
    ----------
    df : pandas.DataFrame
        Data frame to be plotted. A date column is required.
    date_column : str
        Column name of the date in the data frame.
    normal_columns : list of str
        A list of the column names of targets for which inconspicuous
        colors are set.
    stands_out_columns : list of str
        A list of column names for which prominent colors are set.

    Returns
    -------
    legend_dataset : list of dicts
        A list of dictionaries containing data for the legend.
        The following keys are set in the dictionary.
        - key : str -> Column names excluding dates are set.
        - value : The value of the last date is set.
    """
    df.sort_values(by=date_column, inplace=True)
    df.reset_index(drop=True, inplace=True)
    df_len = len(df)
    legend_dataset = []
    merged_columns = [*normal_columns, *stands_out_columns]
    for column_name in merged_columns:
        last_date_val = df.loc[df_len - 1, column_name]
        legend_dataset.append({
            'key': column_name,
            'value': last_date_val,
        })
    return legend_dataset


def _validate_df_columns(
        df, date_column, normal_columns, stands_out_columns):
    """
    Check that the specified column exists in the data frame.

    Parameters
    ----------
    df : pandas.DataFrame
        Data frame to be checked.
        df, date_column, normal_columns, stands_out_columns):
    date_column : str
        Column name of the date in the data frame.
    normal_columns : list of str
        A list of the column names of targets for which inconspicuous
        colors are set.
    stands_out_columns : list of str
        A list of column names for which prominent colors are set.

    Raises
    ------
    ValueError
        If the required column is not included in the data frame.
    """
    has_column = date_column in df.columns
    if not has_column:
        err_msg = 'The specified date column is not included in the data frame.'
        raise ValueError(err_msg)
    merged_column_list = [*normal_columns, *stands_out_columns]
    for column_name in merged_column_list:
        has_column = column_name in df.columns
        if not has_column:
            err_msg = 'The specified column is not included in the data frame : %s' \
                % column_name
            raise ValueError(err_msg)

File: plot_playground/template/storytelling/simple_line_date_series_plot.js
/*
Python Parameters
-----------------
{svg_id} : str
    SVG elemnt's ID.
{svg_width} : int
    Width of SVG area.
{svg_height} : int
    Height of SVG area.
{svg_background_color} : str
    The background color of the svg.
{svg_margin_left} : int
    The left margin of the SVG area. It is basically set by position
    adjustment on Jupyter.
{outer_margin} : int
    Edge margin of SVG area.
{x_ticks} : int
    Number of steps on the x axis. This number roughly varies depending
    on the value of surplus etc.
{y_ticks} : int
    Number of steps in the y axis. This number roughly varies depending
    on the value of surplus etc.
{y_axis_prefix} : str
    A string to set before the value of the y axis.
    e.g., $.
{y_axis_suffix} : str
    A character string to be set after the value of the y axis.
    e.g., %.
{plot_title} : str
    The title of the plot.
{plot_description} : str
    A description of the plot. It is set under the title.
{dataset} : list of dicts
    Data set to set. The following keys are required in the dictionary.
    - date : A string of dates of the form "%Y-%m-%d".
    - other key : Any other key that stores the value of int or float.
{column_list} : list of str
    A list of the column names of targets for which inconspicuous
    colors are set.
{stands_out_column_list} : list of str
    A list of column names for which prominent colors are set.
{legend_dataset} : list of dicts
    A list containing values ​​to be set in the legend. It is used to
    calculate the Y coordinate. Specify a dictionary containing the
    key names key and value. Set the column name to key and value to
    the last value of time series.
    e.g., [{"key": "A", "value": 100}]
{year_str_list} : list of str
    A list containing characters at the beginning of the year.
    e.g., ["2018-01-01", "2019-01-01"]
{y_axis_min} : int or float
    The minimum value on the y axis. Set to 0 or a negative value.
{y_axis_max} : int or float
    Maximum value of y axis.
{y_axis_label} : str
    The label to set on the y axis. It is displayed in a rotated state.
{x_axis_min} : str
    String with the oldest date on the x axis.
{x_axis_max} : str
    String of the last date on the x axis.
 */

const SVG_ID = "{svg_id}";
const SVG_WIDTH = {svg_width};
const SVG_HEIGHT = {svg_height};
const SVG_BACKGROUND_COLOR = "{svg_background_color}";
const SVG_MARGIN_LEFT = {svg_margin_left};
const OUTER_MARGIN = {outer_margin};
const X_TICKS = {x_ticks};
const Y_TICKS = {y_ticks};
const Y_AXIS_PREFIX = "{y_axis_prefix}";
const Y_AXIS_SUFFIX = "{y_axis_suffix}";
const PLOT_TITLE_TXT = "{plot_title}";
const PLOT_DESCRIPTION_TXT = "{plot_description}";

var dateParse = d3.timeParse("%Y-%m-%d");
var dataset = {dataset};
for (var i = 0; i < dataset.length; i++) {
    var dataDict = dataset[i];
    dataDict["date"] = dateParse(dataDict["date"]);
}
const COLUMN_LIST = {column_list};
const STANDS_OUT_COLUMN_LIST = {stands_out_column_list};
var MERGED_COLUMN_LIST = COLUMN_LIST.concat(STANDS_OUT_COLUMN_LIST);
const LEGEND_DATASET = {legend_dataset};
const LEGEND_KEY = function(d) {
    return d.key;
}
var yearDataset = {year_str_list};
for (var i = 0; i < yearDataset.length; i++) {
    yearDataset[i] = dateParse(yearDataset[i]);
}
const Y_AXIS_MIN = {y_axis_min};
const Y_AXIS_MAX = {y_axis_max} * 1.1;
const Y_AXIS_LABEL = "{y_axis_label}";
const X_AXIS_MIN = dateParse("{x_axis_min}");
const X_AXIS_MAX = dateParse("{x_axis_max}");

var svg = d3.select("#" + SVG_ID)
    .style("background-color", SVG_BACKGROUND_COLOR)
    .style("margin-left", SVG_MARGIN_LEFT);

var plotBaseLineY = 0;
if (PLOT_TITLE_TXT !== "") {
    var plotTitle = svg.append("text")
        .attr("x", OUTER_MARGIN)
        .attr("y", OUTER_MARGIN)
        .attr("dominant-baseline", "hanging")
        .text(PLOT_TITLE_TXT)
        .classed("title font", true);
    var plotTitleBBox = plotTitle.node().getBBox();
    plotBaseLineY += plotTitleBBox.y + plotTitleBBox.height;
}

if (PLOT_DESCRIPTION_TXT !== "") {
    var plotDescription = svg.append("text")
        .attr("x", OUTER_MARGIN)
        .attr("y", plotBaseLineY + 10)
        .attr("dominant-baseline", "hanging")
        .text(PLOT_DESCRIPTION_TXT)
        .classed("description font", true);
    var plotDesciptionBBox = plotDescription.node().getBBox();
    plotBaseLineY += plotDesciptionBBox.height + 10;
}

var legend = svg.selectAll("legend")
    .data(LEGEND_DATASET, LEGEND_KEY)
    .enter()
    .append("text")
    .text(function(d) {
        return d.key;
    })
    .attr("dominant-baseline", "central");
legend.each(function(d) {
    var className;
    if (STANDS_OUT_COLUMN_LIST.indexOf(d.key) >= 0) {
        className = "legend stands-out-legend font";
    }else {
        className = "legend font";
    }
    d3.select(this)
        .classed(className, true);
})

var yLabelMarginAdjust = 0;
if (Y_AXIS_LABEL !== "") {
    var yAxisLabel = svg.append("text")
        .text(Y_AXIS_LABEL)
        .attr("transform", "rotate(270)")
        .attr("text-anchor", "end")
        .attr("dominant-baseline", "text-before-edge")
        .classed("font y-axis-label", true);
    yAxisLabel.attr("x", -plotBaseLineY - OUTER_MARGIN + 1)
        .attr("y", OUTER_MARGIN - 3);
    var yAxisLabelBBox = yAxisLabel.node()
        .getBBox();
    yLabelMarginAdjust = yAxisLabelBBox.height + 2;
}
var yAxisScale = d3.scaleLinear()
    .domain([Y_AXIS_MIN, Y_AXIS_MAX])
    .range([SVG_HEIGHT - OUTER_MARGIN, plotBaseLineY + OUTER_MARGIN]);
var yAxis = d3.axisLeft()
    .scale(yAxisScale)
    .ticks(Y_TICKS)
    .tickFormat(function (d) {
        var tickFormat = d;
        if (Y_AXIS_PREFIX !== "") {
            tickFormat = Y_AXIS_PREFIX + tickFormat;
        }
        if (Y_AXIS_SUFFIX !== "") {
            tickFormat += Y_AXIS_SUFFIX;
        }
        return tickFormat;
    });
var yAxisGroup = svg.append("g")
    .classed("y-axis font", true)
    .call(yAxis);
var yAxisBBox = yAxisGroup
    .node()
    .getBBox();
var yAxisPositionX = OUTER_MARGIN + yAxisBBox.width + yLabelMarginAdjust;
yAxisGroup.attr("transform", "translate(" + yAxisPositionX + ", 0)");

var xAxisScale = d3.scaleTime()
    .domain([X_AXIS_MIN, X_AXIS_MAX])
    .range([yAxisPositionX, SVG_WIDTH - OUTER_MARGIN]);

var yearFormat = d3.timeFormat("%Y");
var year = svg.selectAll("year")
    .data(yearDataset)
    .enter()
    .append("text")
    .text(function(d) {
        return yearFormat(d);
    })
    .attr("text-anchor", "middle")
    .attr("x", function(d) {
        return xAxisScale(d);
    })
    .attr("y", SVG_HEIGHT - OUTER_MARGIN)
    .classed("font x-axis-year", true);
var yearBBox = year.node()
    .getBBox()

var xAxis = d3.axisBottom()
    .scale(xAxisScale)
    .ticks(X_TICKS)
    .tickFormat(d3.timeFormat("%m/%d"));
var xAxisGroup = svg.append("g")
    .classed("x-axis font", true)
    .call(xAxis)
var xAxisBBox = xAxisGroup
    .node()
    .getBBox();
xAxisPositionY = parseInt(
    SVG_HEIGHT - OUTER_MARGIN - xAxisBBox.height - yearBBox.height);
xAxisGroup.attr(
    "transform",
    "translate(0, " + xAxisPositionY + ")");

yAxisScale.range([xAxisPositionY, plotBaseLineY + OUTER_MARGIN]);
yAxis.scale(yAxisScale);
yAxisGroup.call(yAxis);

var legendMaxWidth = 0;
svg.selectAll(".legend").each(function(d) {
    var width = d3.select(this)
        .node()
        .getBBox()["width"];
    legendMaxWidth = Math.max(legendMaxWidth, width);
});
svg.selectAll(".legend")
    .attr("x", function(d) {
        return SVG_WIDTH - OUTER_MARGIN - legendMaxWidth;
    })
    .attr("y", function(d) {
        return yAxisScale(d.value);
    });
xAxisScale.range(
    [yAxisPositionX, SVG_WIDTH - OUTER_MARGIN - legendMaxWidth - 10]);
xAxis.scale(xAxisScale);
xAxisGroup.call(xAxis);
year.attr("x", function(d) {
    return xAxisScale(d);
});

var lineGroup = svg.append("g")
    .attr("id", "{svg_id}-lines");
for (var i = 0; i < MERGED_COLUMN_LIST.length; i++) {
    var columnName = MERGED_COLUMN_LIST[i];
    var line = d3.line()
        .x(function (d) {
            return xAxisScale(d.date);
        })
        .y(function (d) {
            return yAxisScale(d[columnName]);
        });
    if (STANDS_OUT_COLUMN_LIST.indexOf(columnName) >= 0) {
        className = "stands-out-line";
    }else {
        className = "line";
    }
    lineGroup.append("path")
        .datum(dataset)
        .classed(className, true)
        .attr("d", line);
}

var xAxisScaleRange = xAxisScale.range();
var yAxisScaleRange = yAxisScale.range();
svg.append("clipPath")
    .attr("id", "{svg_id}-plotAreaClipPath")
    .append("rect")
    .attr("x", xAxisScaleRange[0] + 1)
    .attr("y", yAxisScaleRange[1])
    .attr("width", xAxisScaleRange[1] - xAxisScaleRange[0])
    .attr("height", yAxisScaleRange[0] - yAxisScaleRange[1] - 1);
d3.select("#{svg_id}-lines")
    .attr("clip-path", "url(#{svg_id}-plotAreaClipPath)");
